control-plane/connect-inject/webhook_v2/mesh_webhook.go

```diff
@@ -0,0 +1,580 @@
+	etcResolvFile string
+}
+
+// Handle is the admission.Webhook implementation that actually handles the
+// webhook request for admission control. This should be registered or
+// served via the controller runtime manager.
+func (w *MeshWebhook) Handle(ctx context.Context, req admission.Request) admission.Response {
+	var pod corev1.Pod
+
+	// Decode the pod from the request
+	if err := w.decoder.Decode(req, &pod); err != nil {
+		w.Log.Error(err, "could not unmarshal request to pod")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Marshall the contents of the pod that was received. This is compared with the
+	// marshalled contents of the pod after it has been updated to create the jsonpatch.
+	origPodJson, err := json.Marshal(pod)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Setup the default annotation values that are used for the container.
+	// This MUST be done before shouldInject is called since that function
+	// uses these annotations.
+	if err := w.defaultAnnotations(&pod, string(origPodJson)); err != nil {
+		w.Log.Error(err, "error creating default annotations", "request name", req.Name)
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("error creating default annotations: %s", err))
+	}
+
+	// Check if we should inject, for example we don't inject in the
+	// system namespaces.
+	if shouldInject, err := w.shouldInject(pod, req.Namespace); err != nil {
+		w.Log.Error(err, "error checking if should inject", "request name", req.Name)
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("error checking if should inject: %s", err))
+	} else if !shouldInject {
+		return admission.Allowed(fmt.Sprintf("%s %s does not require injection", pod.Kind, pod.Name))
+	}
+
+	w.Log.Info("received pod", "name", req.Name, "ns", req.Namespace)
+
+	// Add our volume that will be shared by the init container and
+	// the sidecar for passing data in the pod.
+	pod.Spec.Volumes = append(pod.Spec.Volumes, w.containerVolume())
+
+	// Optionally mount data volume to other containers
+	w.injectVolumeMount(pod)
+
+	// Optionally add any volumes that are to be used by the envoy sidecar.
+	if _, ok := pod.Annotations[constants.AnnotationConsulSidecarUserVolume]; ok {
+		var userVolumes []corev1.Volume
+		err := json.Unmarshal([]byte(pod.Annotations[constants.AnnotationConsulSidecarUserVolume]), &userVolumes)
+		if err != nil {
+			return admission.Errored(http.StatusInternalServerError, fmt.Errorf("error unmarshalling sidecar user volumes: %s", err))
+		}
+		pod.Spec.Volumes = append(pod.Spec.Volumes, userVolumes...)
+	}
+
+	// Add the upstream services as environment variables for easy
+	// service discovery.
+	containerEnvVars := w.containerEnvVars(pod)
+	for i := range pod.Spec.InitContainers {
+		pod.Spec.InitContainers[i].Env = append(pod.Spec.InitContainers[i].Env, containerEnvVars...)
+	}
```

Comment:
Maybe we comment this out and add a `TODO` until it works as planned in a future release?

Reply:
added a todo in the Webhook.containerEnvVars function.